site_tools/image_deployer.py

```diff
@@ -0,0 +1,316 @@
+import argparse
+import datetime
+import json
+import logging
+import tempfile
+import yaml
+
+import openstack
+
+from itertools import islice
+
+from swiftclient.client import Connection as SwiftConnection
+
+
+logging.basicConfig(level=logging.INFO)
+
+
+class Image:
+    def __init__(self, name, base_container, scope, current_path):
+        self.name = name
+        self.manifest_name = name + ".manifest"
+        self.raw_name = name + ".raw"
+        self.qcow_name = name + ".qcow2"
+        self.scope = scope
+        self.base_container = base_container
+        self.current_path = current_path
+        self.container_path = self.base_container + "/" + self.current_path
+
+    def __str__(self):
+        return f"Image(name={self.name})"
+
+
+def get_openstack_connection(cloud, cloud_name):
+    return openstack.connect(auth_url=cloud['clouds'][cloud_name]["auth"]["auth_url"],
+                             auth_type="v3applicationcredential",
+                             application_credential_id=cloud['clouds'][cloud_name]["auth"]["application_credential_id"],
+                             application_credential_secret=cloud['clouds'][cloud_name]["auth"]["application_credential_secret"],
+                             region_name=cloud['clouds'][cloud_name]["region_name"])
+
+
+def get_current_value(connection, base_container, scope):
+    _, current = connection.get_object(base_container, scope + "/current")
+    return current.strip()
+
+
+def get_available_images(connection, base_container, scope, current):
+    logging.info("Checking available images...")
+    available_images = []
+
+    current_path = scope + "/" + current
+    current_objects = connection.object_store.objects(base_container,
+                                                      prefix=current_path,
+                                                      delimiter="/")
+
+    for obj in islice(current_objects, 1, None):
+        object_name = obj.name.split("/")[-1]
+        logging.debug(f"Checking object: {object_name}")
+        if object_name.endswith(".manifest"):
+            name = object_name.rstrip(".manifest")
+            available_images.append(Image(name, base_container, scope, current_path))
+
+    return available_images
+
+
+def get_site_images(connection):
+    logging.info("Checking site images...")
+    images = [i.name for i in connection.image.images()]
```

Comment:
an optimization to make, with an admin credential this will list all images owned by all projects, a very long list.

all "published" images would have `public` visibility, and we can filter for that here.

Reply:
added an option to filter based on this, but again, mostly left the old way for my own testing. we can remove it in the final version